README.md
Description
===========

Installs and configures Graphite (http://graphite.readthedocs.org)

Requirements
============

* CentOS 6
* Red Hat Enterprise Linux 6
* Fedora

Note: CentOS/RHEL 5 is too old to run Graphite on. In particular, its lack of
sufficiently new Twisted libraries hinders it.

Attributes
==========

* `node[:graphite][:server][:password]` sets the default password for the
   graphite "root" user.
* `node[:graphite][:web][:virtualhostname]` sets the virtual host name under
   which you are running the Graphite Composer and render API.

Usage
=====

`recipe[graphite::server]` should build a stand-alone Graphite installation.

Caveats
=======

Ships with two default schemas, stats.* (for Etsy's statsd) and a
catchall that matches anything. The catchall retains minutely data for
13 months, as in the default config. stats retains data every 10 seconds
for 6 hours, every minute for a week, and every 10 minutes for 5 years.

Todo
====

Fix test_kitchen support (not sure that the tests actually work)

Credits
=======

Forked from Heavy Water Software: https://github.com/heavywater/chef-graphite


CHANGELOG.md
## 1.0.0

- Initial release to GitHub


recipes/carbon.rb
#
# Cookbook Name:: graphite
# Recipe:: default
#
# Copyright 2012, SecondMarket Labs, LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations unde

include_recipe "python"

package "python-twisted-core" do
  action :install
end

python_pip "carbon" do
  action :install
  version node[:graphite][:version]
end

template "/etc/init.d/carbon-cache" do
  owner "root"
  group "root"
  mode 00755
  source "carbon-cache.init.erb"
end

user node[:graphite][:carbon][:user] do
  action :create
end

service "carbon-cache" do
  supports [ :start, :stop, :restart ]
  action :enable
end

template "/opt/graphite/conf/carbon.conf" do
  owner "root"
  group "root"
  mode 00644
  variables( :line_receiver_interface => node[:graphite][:carbon][:line_receiver_interface],
             :pickle_receiver_interface => node[:graphite][:carbon][:pickle_receiver_interface],
             :cache_query_interface => node[:graphite][:carbon][:cache_query_interface] )
  notifies :restart, "service[carbon-cache]"
end

template "/opt/graphite/conf/storage-schemas.conf" do
  owner "root"
  group "root"
  mode 00644
end

%w{/opt/graphite/storage/whisper /opt/graphite/storage/log/carbon-cache}.each do |d|
  directory d do
    owner node[:graphite][:carbon][:user]
    group node[:graphite][:carbon][:group]
    action :create
    mode 00755
  end
end

service "carbon-cache" do
  action :start
end


attributes/graphite.rb
#
# Cookbook Name:: graphite
# Attributes:: default
# Author:: Julian C. Dunn <jdunn@secondmarket.com>
#
# Copyright 2012, SecondMarket Labs, LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

default[:graphite][:version] = "0.9.10"

default[:graphite][:carbon][:uri] = "http://launchpadlibrarian.net/82112362/carbon-#{node[:graphite][:version]}.tar.gz"
default[:graphite][:carbon][:checksum] = "b3d42e3b93c09"

default[:graphite][:whisper][:uri] = "http://launchpadlibrarian.net/82112367/whisper-#{node[:graphite][:version]}.tar.gz"
default[:graphite][:whisper][:checksum] = "66c05eafe8d86"

default[:graphite][:graphite_web][:uri] = "http://launchpadlibrarian.net/82112308/graphite-web-#{node[:graphite][:version]}.tar.gz"
default[:graphite][:graphite_web][:checksum] = "cc78bab7fb26b"

default[:graphite][:carbon][:line_receiver_interface] =   "0.0.0.0"
default[:graphite][:carbon][:pickle_receiver_interface] = "0.0.0.0"
default[:graphite][:carbon][:cache_query_interface] =     "0.0.0.0"
default[:graphite][:carbon][:user] = "carbon"
default[:graphite][:carbon][:group] = "carbon"

default[:graphite][:web][:password] = "graphite"
default[:graphite][:web][:virtualhostname] = "localhost"


recipes/web.rb
#
# Cookbook Name:: graphite
# Recipe:: web
#
# Copyright 2012, SecondMarket Labs, LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
include_recipe "apache"
include_recipe "apache::mod_wsgi"

%w{pycairo Django django-tagging python-ldap python-twisted-core python-memcached python-sqlite2}.each do |p|
  package p do
    action :install
  end
end

python_pip "graphite-web" do
  action :install
  version node[:graphite][:version]
end

cookbook_file "/opt/graphite/conf/graphite.wsgi" do
  owner "root"
  group "root"
  mode 00755
  action :create
end

apache_virtualhost "http #{node[:graphite][:web][:virtualhostname]}" do
  cookbook "graphite"
  servername node[:graphite][:web][:virtualhostname]
  serveralias node[:fqdn]
  template "graphite-vhost.conf.erb"
end

directory "/opt/graphite/storage" do
  owner node[:apache][:user]
  group node[:apache][:group]
end

directory "/opt/graphite/storage/log/webapp" do
  owner node[:apache][:user]
  group node[:apache][:group]
end

cookbook_file "/opt/graphite/bin/set_admin_passwd.py" do
  owner "root"
  group "root"
  mode 00755
end

cookbook_file "/opt/graphite/storage/graphite.db" do
  action :create_if_missing
  owner node[:apache][:user]
  group node[:apache][:group]
  mode 00644
  notifies :run, "execute[set-admin-password]"
end

execute "set-admin-password" do
  command "/opt/graphite/bin/set_admin_passwd.py root #{node[:graphite][:web][:password]}"
  action :nothing
end
